### ISSUE TITLE
Mismatch in results in paper and in shared weights

### ISSUE THREAD

**Issue by njan-creative** [zhi-xuan-chen/Reg2RG#14]
Evaluated using shared weights - 'Avg. Pr': 0.3985568185257445, 'Avg. Re': 0.17036592681463708, 'Avg. F1': 0.22581017449311405

Paper stats - Precision-0.423 Recall-0.181 F1-0.253

There is a mismatch in inference using weights shared and paper stats.

Evaluated using shared weights - 'Overall_BLEU-1': '0.5435', 'Overall_BLEU-2': '0.4734', 'Overall_BLEU-3': '0.4253', 'Overall_BLEU-4': '0.3855', 'Overall_METEOR': '0.6183', 'Overall_ROUGE-1': '0.6088', 'Overall_ROUGE-2': '0.4797', 'Overall_ROUGE-L': '0.5652'

Paper stats - BLEU-1 : 49.63 BLEU-2 : 41.43 BLEU-3 : 35.91 BLEU-4 : 32.04 METEOR : 49.71 ROUGE-L : 47.76

**Comment by Kanade-hub** [comment]
Did you run rm_region_text.py? After generating the predicted text, you need to run this program to remove the prefix in order to get the real report. However, even after removing the prefix, I still can't reproduce the results from the paper.

**Comment by Jorwnpay** [comment]
I have spent over two weeks attempting to replicate the results presented in this paper, but to no avail. Interestingly, when I completely removed the core innovation—the local regional feature branch—the performance actually improved. I am at a loss for words.

**Comment by zhi-xuan-chen** [comment]
@njan-creative When I do the experiment, I find different inference params will also affect the performance. Do you use the same params and same version of transformer? It seems NLG improves but CE drops in your setting. And we also provide our inferenced result in this repo, do you test it? Whether it can get the same results in our paper?

**Comment by zhi-xuan-chen** [comment]
@Jorwnpay Can you share your detailed setting of this experiemt? In our experiment, we truly find our regional information can help a lot.

**Comment by Jorwnpay** [comment]
> [@Jorwnpay](https://github.com/Jorwnpay) Can you share your detailed setting of this experiemt? In our experiment, we truly find our regional information can help a lot.

Thank you for your prompt response! I would like to provide more details regarding my replication experiments.

For the original Reg2RG experiments, I maintained the default settings provided in this repository and conducted three runs using different random seeds. The experiments were performed on 2x NVIDIA A800 GPUs. The specific hyperparameters are as follows:

```
# Training settings
learning_rate=5e-5
per_device_train_batch_size=1
num_train_epochs=10
gradient_accumulation_steps=8
evaluation_strategy="no"
save_strategy="epoch"
save_total_limit=3
weight_decay=0.0
warmup_steps=20
lr_scheduler_type="constant_with_warmup"
dataloader_num_workers=8
dataloader_pin_memory=False 
logging_steps=1
```

The DeepSpeed configuration (`dsconfig`) is as follows:

```
{
  "train_micro_batch_size_per_gpu": "auto",
  "gradient_accumulation_steps": "auto",
  "gradient_clipping": "auto",
  "zero_allow_untested_optimizer": true,
  "bf16": {
    "enabled": "auto"
  },  
  "zero_optimization": {
    "stage": 2,
    "allgather_partitions": true,
    "allgather_bucket_size": 1e9,
    "reduce_scatter": true,
    "reduce_bucket_size": 1e9,
    "overlap_comm": true,
    "contiguous_gradients": true
  }
}
```

The experimental results are summarized in the table below. **Reg2RG (Original)** refers to the results reported in the paper; **Reg2RG (Seed=4x)** represents my replication results with different seeds; and **Reg2RG (Seed=42, w/o regional branch)** denotes the ablation study where the local regional feature branch was removed, keeping only the global branch.

| **Methods**                           | **BLEU-1** | **BLEU-2** | **BLEU-3** | **BLEU-4** | **METEOR** | **ROUGE-L** | **Precision** | **Recall** | **F1**   |
| ------------------------------------- | ---------- | ---------- | ---------- | ---------- | ---------- | ----------- | ------------- | ---------- | -------- |
| Reg2RG (Original)                     | **47.25**  | 36.49      | 29.57      | 24.87      | **44.07**  | 36.65       | **0.42**      | **0.18**   | **0.25** |
| Reg2RG (Seed=41)                      | 45.86      | 35.27      | 28.46      | 23.96      | 42.64      | 36.74       | 0.352         | 0.150      | 0.210    |
| Reg2RG (Seed=42)                      | 45.50      | 34.72      | 27.73      | 23.13      | 42.19      | 35.98       | 0.396         | 0.155      | 0.223    |
| Reg2RG (Seed=42, w/o regional branch) | 44.24      | 34.81      | 28.59      | 24.34      | 40.31      | 32.15       | 0.415         | 0.163      | 0.230    |
| Reg2RG (Seed=43)                      | 46.89      | **36.77**  | **30.35**  | **26.09**  | 43.79      | **37.64**   | 0.325         | 0.107      | 0.160    |

**Observations:**

1. **Seed Sensitivity:** There appears to be a trade-off between NLG and CE metrics across different random initializations. For instance, while Seed=42 yields the best CE performance among the three, its NLG metrics are lower. Conversely, Seed=43 shows slightly better NLG performance than the original paper but suffers a 9% drop in the CE metric.
2. **Ablation Anomaly:** When removing the regional branch (Seed=42), although METEOR and ROUGE-L decreased, the BLEU scores were slightly higher than the full model with the same seed, and all CE-related metrics actually showed improvement.

While the NLG metrics are generally reproducible, I am concerned about the consistency of the regional branch's contribution across different environments. Could these fluctuations be attributed to hardware differences (e.g., A800 vs. 4090) or perhaps specific environment dependencies?

I would appreciate any insights you might have on how to stabilize these results. Thank you again for your contribution to the community!

**Comment by njan-creative** [comment]
@Jorwnpay Is this inference results with shared weights or training results?
For me CE was similar but not NLG metrics.

**Comment by Jorwnpay** [comment]
> [@Jorwnpay](https://github.com/Jorwnpay) Is this inference results with shared weights or training results? For me CE was similar but not NLG metrics.

This is the results of my retrained model through the shared codes. As for the shared weights, I can totally reproduce both the NLG and CE metrics.

**Comment by njan-creative** [comment]
@Jorwnpay How did you fix this issue.

https://github.com/zhi-xuan-chen/Reg2RG/issues/13
https://github.com/zhi-xuan-chen/Reg2RG/issues/8

Did you use strict=False as below ?
model.load_state_dict(torch.load(model_path, map_location=device), strict=False)

**Comment by Kanade-hub** [comment]
> > [@Jorwnpay](https://github.com/Jorwnpay) Is this inference results with shared weights or training results? For me CE was similar but not NLG metrics.
> 
> This is the results of my retrained model through the shared codes. As for the shared weights, I can totally reproduce both the NLG and CE metrics.

Are you using the evaluation method from the code in this repository? I used the evaluation method from the pycocoevalcap library with shared weights for inference, and the results I obtained were only blue1-41.50, blue2-31.81, blue3-26.10, and blue4-22.18. The experimental environment is completely consistent with the paper.

**Comment by zmmdeid** [comment]
> > [@Jorwnpay](https://github.com/Jorwnpay) 这种推断是基于权重共享还是训练结果？我自己也差不多，但不是 NLG 的指标。
> 
> 这是我通过共享代码重新训练模型的结果。至于共享权重，我完全能复现 NLG 和 CE 的指标。



> > [@Jorwnpay](https://github.com/Jorwnpay) Can you share your detailed setting of this experiemt? In our experiment, we truly find our regional information can help a lot.
> 
> Thank you for your prompt response! I would like to provide more details regarding my replication experiments.
> 
> For the original Reg2RG experiments, I maintained the default settings provided in this repository and conducted three runs using different random seeds. The experiments were performed on 2x NVIDIA A800 GPUs. The specific hyperparameters are as follows:
> 
> ```
> # Training settings
> learning_rate=5e-5
> per_device_train_batch_size=1
> num_train_epochs=10
> gradient_accumulation_steps=8
> evaluation_strategy="no"
> save_strategy="epoch"
> save_total_limit=3
> weight_decay=0.0
> warmup_steps=20
> lr_scheduler_type="constant_with_warmup"
> dataloader_num_workers=8
> dataloader_pin_memory=False 
> logging_steps=1
> ```
> 
> The DeepSpeed configuration (`dsconfig`) is as follows:
> 
> ```
> {
>   "train_micro_batch_size_per_gpu": "auto",
>   "gradient_accumulation_steps": "auto",
>   "gradient_clipping": "auto",
>   "zero_allow_untested_optimizer": true,
>   "bf16": {
>     "enabled": "auto"
>   },  
>   "zero_optimization": {
>     "stage": 2,
>     "allgather_partitions": true,
>     "allgather_bucket_size": 1e9,
>     "reduce_scatter": true,
>     "reduce_bucket_size": 1e9,
>     "overlap_comm": true,
>     "contiguous_gradients": true
>   }
> }
> ```
> 
> The experimental results are summarized in the table below. **Reg2RG (Original)** refers to the results reported in the paper; **Reg2RG (Seed=4x)** represents my replication results with different seeds; and **Reg2RG (Seed=42, w/o regional branch)** denotes the ablation study where the local regional feature branch was removed, keeping only the global branch.
> 
> **Methods**	**BLEU-1**	**BLEU-2**	**BLEU-3**	**BLEU-4**	**METEOR**	**ROUGE-L**	**Precision**	**Recall**	**F1**
> Reg2RG (Original)	**47.25**	36.49	29.57	24.87	**44.07**	36.65	**0.42**	**0.18**	**0.25**
> Reg2RG (Seed=41)	45.86	35.27	28.46	23.96	42.64	36.74	0.352	0.150	0.210
> Reg2RG (Seed=42)	45.50	34.72	27.73	23.13	42.19	35.98	0.396	0.155	0.223
> Reg2RG (Seed=42, w/o regional branch)	44.24	34.81	28.59	24.34	40.31	32.15	0.415	0.163	0.230
> Reg2RG (Seed=43)	46.89	**36.77**	**30.35**	**26.09**	43.79	**37.64**	0.325	0.107	0.160
> **Observations:**
> 
> 1. **Seed Sensitivity:** There appears to be a trade-off between NLG and CE metrics across different random initializations. For instance, while Seed=42 yields the best CE performance among the three, its NLG metrics are lower. Conversely, Seed=43 shows slightly better NLG performance than the original paper but suffers a 9% drop in the CE metric.
> 2. **Ablation Anomaly:** When removing the regional branch (Seed=42), although METEOR and ROUGE-L decreased, the BLEU scores were slightly higher than the full model with the same seed, and all CE-related metrics actually showed improvement.
> 
> While the NLG metrics are generally reproducible, I am concerned about the consistency of the regional branch's contribution across different environments. Could these fluctuations be attributed to hardware differences (e.g., A800 vs. 4090) or perhaps specific environment dependencies?
> 
> I would appreciate any insights you might have on how to stabilize these results. Thank you again for your contribution to the community!

Hello, I also trained and inferred using the author's parameters, and my results are roughly the same as yours. I've tried twice, but the values ​​are similar to those in your table. I've also tried using other parameters, but the results are about the same. Did you eventually manage to reproduce the result? Did the metrics change?
